src/Transport/Sending/NativeMessageCustomizationBehavior.cs

```diff
@@ -0,0 +1,48 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.ServiceBus;
+    using Pipeline;
+    using Transport;
+
+    class NativeMessageCustomizationBehavior : Behavior<IRoutingContext>
+    {
+        readonly bool isOutboxEnabled;
+
+        public NativeMessageCustomizationBehavior(bool isOutboxEnabled)
+        {
+            this.isOutboxEnabled = isOutboxEnabled;
+        }
+
+        internal const string CustomizationKey = "$ASB.CustomizationId";
+
+        public override Task Invoke(IRoutingContext context, Func<Task> next)
+        {
+            if (!context.Extensions.TryGet(CustomizationKey, out Action<Message> customization))
+            {
+                return next();
+            }
+
+            if (isOutboxEnabled)
+            {
+                throw new Exception("Native message customization cannot be used together with the Outbox as customizations are not persistent. Disable the outbox to use native message customization.");
+            }
+
+            // When part of an incoming message, the transport transaction is set by the transport.
+            // Otherwise it will be created at this point which works because there is no batched dispatch for IMessageSession operations.
+            var transportTransaction = context.Extensions.GetOrCreate<TransportTransaction>();
+
+            // Use the TransportTransaction to store complex objects and pass them to the transport. The transaction might be shared by multiple send operations.
+            var customizer = transportTransaction.GetOrCreate<NativeMessageCustomizer>();
+
+            var customizationId = Guid.NewGuid().ToString();
+            customizer.Customizations.TryAdd(customizationId, customization);
+
+            // Store they key to the customization in the dispatch properties that are passed to the transport
+            context.Extensions.Get<DispatchProperties>()[CustomizationKey] = customizationId;
```

Comment:
> Could we store both customization and key only in DispatchProperties or TransportTransaction?

we can't store the customization in the properties as it can't be serialized to a string.
we can't store the key in the transaction as the transaction might be shared across multiple send operations and there is not easy way to find the matching customization of a specific operation again.

> Could we store only customization without the key as we support at most one customization per Send operation?

but multiple send operations might happen on the same incoming message which share one transaction.